### PR TITLE
feat(magic-graphs): add basic templates

### LINKED ISSUE #231: add graph templates unique to each product for loading in interesting graphs

### PR DESCRIPTION
**Pull request by teykamp**
fixes #231 

- added templates inside of info.ts for each product
- added ability to load templates
- added template menu
- added snapshot api for autogenerating template images when added
- added ability to save current graph as template

### CODE REVIEW

File: client/src/graphs/templates/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import type { GNode } from "@graph/types";
+import type { BoundingBox, Coordinate } from "@shape/types";
+
+export const getAverageCoordinatesOfGraphNodes = (nodes: GNode[]): Coordinate => {
```
Comment (Yonava) [inline client/src/graphs/templates/helpers.ts:4]: id remove implicit returns. and i would get rid of Graph from the title

File: client/src/graphs/templates/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  const total = nodes.reduce((acc, node) => {
+    acc.x += node.x;
+    acc.y += node.y;
+    return acc;
+  }, { x: 0, y: 0 });
+
+  const nodeCount = nodes.length;
+  return {
+    x: total.x / nodeCount,
```
Comment (Yonava) [inline client/src/graphs/templates/helpers.ts:17]: wanna round this?

Comment (Yonava) [reply client/src/graphs/templates/helpers.ts:17]: Look at roundToNearestN Math.util

File: client/src/graphs/templates/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    y: total.y / nodeCount
+  };
+};
+
+export const getGraphNodesBoundingBox = (nodes: GNode[], padding: number = 0): BoundingBox => {
```
Comment (Yonava) [inline client/src/graphs/templates/helpers.ts:22]: this is already done for you in Marquee updateEncapsulatedNodeBox. move your tests to the marquee plugin just as it appears in graph/base (take a look). and then take the function for updateEncapsulatedNodeBox and move it into its adjacent files and then import back to maqree and use it here

File: client/src/graphs/templates/templates.test.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import { describe, it, expect } from "vitest";
+import {
+  getAverageCoordinatesOfGraphNodes,
+  centerNodesOnOriginCoordinates,
+} from "./helpers";
+
+describe("getAverageCoordinatesOfGraphNodes", () => {
+  it("should return the average coordinates for a list of nodes", () => {
```
Comment (Yonava) [inline client/src/graphs/templates/templates.test.ts:8]: no! ❌  ~~it("should return the average coordinates for a list of nodes"~~ 

✅ it("returns the average coordinates of the nodes"

File: client/src/graphs/templates/templates.test.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  });
+
+  it("should handle a single node correctly", () => {
+    const nodes = [{ x: 15, y: 25 }];
+
+    const result = getAverageCoordinatesOfGraphNodes(nodes);
+    expect(result).toEqual({ x: 15, y: 25 });
+  });
+
+  it("should return { x: 0, y: 0 } for an empty node list", () => {
```
Comment (Yonava) [inline client/src/graphs/templates/templates.test.ts:26]: handles edge case of 0 nodes instead of specifically putting the object inside the name

File: client/src/graphs/templates/templates.test.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    const nodes: { x: number; y: number }[] = [];
```
Comment (Yonava) [inline client/src/graphs/templates/templates.test.ts:27]: this is a good place to use the coordinate type

File: client/src/graphs/templates/types.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import type { GEdge, GNode } from "@graph/types";
+
+export type GraphTemplate = {
+  id: string
+  productId?: string
```
Comment (Yonava) [inline client/src/graphs/templates/types.ts:5]: productId can already be inferred based on the product it is originating from

File: client/src/graphs/templates/ui/TemplateItem.vue [hunk 8]
```diff
@@ -0,0 +1,54 @@
+<script setup lang="ts">
+import GButton from "@ui/graph/button/GButton.vue";
+import type { GraphTemplate } from "../types";
+import { ref, toRefs } from "vue";
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateItem.vue:4]: nit: vue is always at the top

File: client/src/graphs/templates/ui/TemplateItem.vue [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import colors from "@utils/colors";
+
+const props = defineProps<{
+  template: GraphTemplate;
+  loadTemplate: (id: string) => void;
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateItem.vue:9]: id call these load and delete if we already know the context its in. however delete is a keyword conflict so maybe remove?

File: client/src/graphs/templates/ui/TemplateItem.vue [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    <div
+      class="absolute w-full h-full z-10 flex items-center justify-center transition duration-200"
+      :style="{
+        opacity: hovered ? 1 : 0,
+      }"
+    >
+      <div class="flex gap-2">
+        <GButton @click="loadTemplate(template.id)">Load</GButton>
+        <GButton
+          v-if="!template.productId"
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateItem.vue:46]: this is meant to be if its user added? If so thats confusing as fuck for a non-informed individual. so make a fucking name in the template that is "isUserAdded"

Comment (Yonava) [reply client/src/graphs/templates/ui/TemplateItem.vue:46]: or make a more sustainable solution then using productId cus what if the user wants to add a template that is specific to a product

File: client/src/graphs/templates/ui/TemplateMenu.vue [hunk 11]
```diff
@@ -0,0 +1,63 @@
+<script setup lang="ts">
+  import { graph } from "@graph/global";
+  import CPopover from "@ui/core/Popover.vue";
+  import GWell from "@ui/graph/GWell.vue";
+  import { useTemplate } from "../useTemplate";
+  import GButton from "@ui/graph/button/GButton.vue";
+  import TemplateItem from "./TemplateItem.vue";
+import { computed } from "vue";
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateMenu.vue:8]: :(

Comment (Yonava) [reply client/src/graphs/templates/ui/TemplateMenu.vue:8]: for two reasons

Comment (Yonava) [reply client/src/graphs/templates/ui/TemplateMenu.vue:8]: two spaces."  "

Comment (Yonava) [reply client/src/graphs/templates/ui/TemplateMenu.vue:8]: tab

Comment (Yonava) [reply client/src/graphs/templates/ui/TemplateMenu.vue:8]: three reasons

File: client/src/graphs/templates/ui/TemplateMenu.vue [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+  const {
+    templates,
+    userTemplates,
+    addCurrentGraphAsTemplate,
+    loadTemplate,
+    clearUserTemplates,
+    deleteUserTemplate,
+  } = useTemplate(graph.value);
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateMenu.vue:17]: why is this not useGraphTemplate? I suggest it to be that cus theres something called useTemplateRef and useTemplate in vue and it can be confusing in the context of a vue SFC

Comment (Yonava) [reply client/src/graphs/templates/ui/TemplateMenu.vue:17]: AND if its useGraphTemplate, you dont have to include the name template in each exported property name

File: client/src/graphs/templates/ui/TemplateMenu.vue [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    </template>
+    <GWell
+      class="flex flex-col text-xl font-bold p-3 w-[400px] rounded-lg gap-2"
+    >
+      <h1 class="text-2xl">Templates</h1>
+      <p
+        v-if="templates.length === 0"
+        class="font-normal text-base text-center"
+      >
+        Add a template!
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateMenu.vue:35]: why are you so happy

File: client/src/graphs/templates/ui/TemplateMenu.vue [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+              backgroundColor: color,
+            }"
+          />
+        </div>
+      </div>
+
+      <hr class="my-2" />
+
+      <div class="flex gap-2 justify-center">
+        <GButton @click="addCurrentGraphAsTemplate" :disabled="graph.nodes.value.length === 0">Save Current</GButton>
```
Comment (Yonava) [inline client/src/graphs/templates/ui/TemplateMenu.vue:58]: one line attrs!

File: client/src/graphs/templates/useTemplate.ts [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+  getGraphNodesBoundingBox,
+} from "./helpers";
+import { generateId } from "@utils/id";
+import { useLocalStorage } from "@vueuse/core";
+import { createImageFromCanvasRegion } from "./snapshot";
+import { products } from "@utils/product";
+
+export const useTemplate = (graph: Graph) => {
+  const userTemplates = useLocalStorage<GraphTemplate[]>(
+    "user-graph-templates",
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:16]: i dont think you need to prepend this with "user-" if its going into local storage

File: client/src/graphs/templates/useTemplate.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    products.flatMap((p) => p.templates ?? [])
+  );
+
+  const templates = computed(() => [
+    ...userTemplates.value, // user templates first so easier to find
+    ...productTemplates.value,
+  ]);
+
+  const addCurrentGraphAsTemplate = (
+    templateDetails: Pick<GraphTemplate, "title" | "description" | "thumbnail">
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:29]: id just call this options. thats what we are calling them everywhere🔒

File: client/src/graphs/templates/useTemplate.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (!graph.canvas.value) return;
+
+    const thumbnail = createImageFromCanvasRegion(
+      graph.canvas.value,
+      boundingBox
+    );
+
+    // console.log({
+    //   nodes: JSON.parse(JSON.stringify(nodes.value)),
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:48]: ✨

File: client/src/graphs/templates/useTemplate.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    const { nodes, edges } = graph;
+
+    const nodeRadius = graph.baseTheme.value.nodeSize;
+    const nodeBorderWidth = graph.baseTheme.value.nodeBorderWidth;
+    const boundingBox = getGraphNodesBoundingBox(
+      nodes.value,
+      nodeRadius + nodeBorderWidth / 2
+    );
+
+    if (!graph.canvas.value) return;
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:40]: its kinda weird to do this here. either make it guard at the top or not at all, we already do a bunch of canvas computations. i would also just edit the param of image to take it a ref that could be undefined. GetCtx does the checks for you

File: client/src/graphs/templates/useTemplate.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      graphState: {
+        nodes: JSON.parse(JSON.stringify(nodes.value)),
+        edges: JSON.parse(JSON.stringify(edges.value)),
+      },
+    });
+  };
+
+  const loadTemplate = (templateId: GraphTemplate["id"]) => {
+    const template = templates.value.find((t) => t.id === templateId);
+    if (template === undefined) {
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:66]: you can just do !template here

File: client/src/graphs/templates/useTemplate.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+      throw new Error(`template could not be loaded: ${templateId} not found`);
+    }
+
+    const { nodes, edges } = template.graphState;
+    const { x, y } = getAverageCoordinatesOfGraphNodes(graph.nodes.value);
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:71]: ⛓️ just say coords cus you are never using them individually.

File: client/src/graphs/templates/useTemplate.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+  const deleteUserTemplate = (templateId: GraphTemplate["id"]) => {
+    userTemplates.value = userTemplates.value.filter(
+      (t) => t.id !== templateId
+    );
+  };
+
+  return {
+    addCurrentGraphAsTemplate,
+    loadTemplate,
+    clearUserTemplates,
```
Comment (Yonava) [inline client/src/graphs/templates/useTemplate.ts:89]: get rid of the redundant naming here. call it add, load, remove

File: client/src/products/dijkstras/templates.ts [hunk 22]
```diff
@@ -0,0 +1,181 @@
+import { generateId } from "@utils/id";
+
+const templates = [
+  {
+    id: generateId(),
```
Comment (Yonava) [inline client/src/products/dijkstras/templates.ts:5]: if its inside the binary we ship, you should give this a descriptive id name

File: client/src/products/dijkstras/templates.ts [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+    productId: 'dijkstras',
+    title: 'Dijkstras',
+    description: 'basic example for visualizing Dijkstras Algorithm',
```
Comment (Yonava) [inline client/src/products/dijkstras/templates.ts:8]: im giving you shit for this
